src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java

```diff
@@ -0,0 +1,49 @@
+package org.kairosdb.core.http.rest;
+
+import com.google.common.collect.SetMultimap;
+import org.kairosdb.core.datastore.QueryMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class QueryAutocompleter {
+    private static final Logger logger = LoggerFactory.getLogger(QueryAutocompleter.class);
+
+    public void complete(QueryMetric query) {
+        completeMetricTag(query);
+    }
+
+    private void completeMetricTag(QueryMetric query) {
+        final SetMultimap<String, String> tags = query.getTags();
+        final Set<String> keys = tags.get("key");
+        final Set<String> metrics = new HashSet<>();
+        for (final String key : keys) {
+            try {
+                final String metric = extractMetricName(key);
+                if (isWildcard(metric)) {
+                    return;
+                }
+                metrics.add(metric);
+            } catch (Exception e) {
+                logger.warn("Problem while parsing key: " + key, e);
+                return;
+            }
+        }
+        tags.putAll("metric", metrics);
+    }
+
+    private String extractMetricName(final String key) {
+        if (null == key || "".equals(key)) return null;
+        final String[] keyParts = key.split("\\.");
+        final String metricName = keyParts[keyParts.length - 1];
+        return "".equals(metricName) ? keyParts[keyParts.length - 2] : metricName;
```

Comment:
why `keyParts[keyParts.length - 2]`

Reply:
This piece of logic was copied [from data-service](https://github.com/zalando-zmon/zmon-data-service/blob/master/src/main/java/de/zalando/zmon/dataservice/data/KairosDBStore.java#L103-L111). I assume there are some (rather rare) cases when key ends with `.`, and then `keyParts.length - 1` is an empty string and and we should use the second from the end section as a `metric`.

Reply:
Just FYI: Java's `String.split(regex)` ignores trailing empty strings.
So `aaa.bbb.` will be split to `["aaa", "bbb"]`